guide/contributing/codebase-overview.md

```diff
@@ -80,28 +84,68 @@ npm run scaffold [ComponentName]
 
 This will output a set of files in the aforementioned structure.
 
-### React & JSX patterns
+### React & JSX practices
 
-Though the following patterns are not strictly enforced, they are strongly encouraged:
+Though the following practices are not strictly enforced, they are strongly encouraged:
+
+### Code style and conventions
+
+Here are some dos and don'ts to consider when writing code.
 
 #### General
 
-- Prefer functional stateless components. Only create a class when you either need state or need lifecycle methods
+- When creating components, consider this order of preference:
+  - Functional components, optionally with hooks and effects
+  - Pure components
+  - Class components
 - Accessibility and responsiveness is a must for every component
 - Provide a flexible and predictable interface via props
+- Prefer containment over specialization. See [React docs on composition vs inheritance][react-composition]
 - Components should only be aware of themselves or their direct children
+- Reuse component logic and shared assets as much as possible
+  - Give particular consideration to [@tds/core-colours][tds-colours], [@tds/shared-styles][tds-styles], [@tds/shared-typography][tds-type], and other core components
+
+#### React
 
-#### Props
+- **DO** make components self-contained. A component should not know anything about other components, except for its direct children
+- **DO** provide a clear, prop-based API to the component. Avoid allowing consumers to customize styles by passing in `className` or `style` as this is not a clear API. Use a prop with known values instead
+- **DO** use React component state for ephemeral UI state within components, while avoiding redux or other application state containers
+- **DO** use dependencies when needed, such as lodash functions or other open source React components
```

Comment:
Not sure if we should advise on using `lodash`. In most cases you can write your own utility or borrow lodash code which will always be more performant than using a lib.

Reply:
I agree. The actual "do" is ok, but let's not use `lodash` as an example.